**Name : Vinayak Renu Nair**

**Panther ID : 002553736**

In [ ]:
!tar -xvf '/content/drive/MyDrive/review_polarity.tar.gz'

**1) Using NLTK tokenize all documents, separated by polarity, remove stop words, and list
the top 20 most frequent tokens (and their counts) for the positive reviews, and the top
20 most frequent tokens (and their counts). What kind of things do you notice are
different between the two sets?** (30 points)

In [2]:
import nltk
nltk.download('stopwords')
nltk.download('punkt')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

In [3]:
from nltk.corpus import stopwords  
from nltk.tokenize import word_tokenize  

In [4]:
import os

In [5]:
stop_words = set(stopwords.words('english'))

In [6]:
path = '/content/txt_sentoken/pos/'

In [7]:
txt = os.listdir(path)

In [8]:
d = []

for f in txt :
  d.append(open(os.path.join(path,f), 'r').read())

In [9]:
len(d)

1000

In [10]:
strn = ''.join(d)
pos_text = word_tokenize(strn)

In [11]:
path1 = '/content/txt_sentoken/neg/'

In [12]:
txt1 = os.listdir(path1)

In [13]:
da = []

for h in txt1 :
  da.append(open(os.path.join(path1,h), 'r').read())

In [14]:
len(da)

1000

In [15]:
strm = ''.join(da)
t = word_tokenize(strm)

Removing Stop words in Negative files

In [16]:
filtered_sentence = [] 

In [17]:
for w in t:  
    if w not in stop_words:  
        filtered_sentence.append(w) 

In [18]:
frequency = {}

for token in filtered_sentence:
  if token in frequency: frequency[token] +=1
  else : frequency[token] = 1

In [19]:
sorted_dictionary = sorted(frequency.items(), key=lambda x:-x[1])[:20]

Top 20 frequent tokens in Negative files

In [93]:
print("Top 20 frequent tokens in Negative files are :")
print("\n")
print(sorted_dictionary)

Top 20 frequent tokens in Negative files are :


[(',', 35269), ('.', 32162), ('``', 9123), ("'s", 8655), (')', 5742), ('(', 5650), ('film', 4257), ("n't", 3442), ('movie', 3174), ('one', 2637), ('?', 2201), ('like', 1832), (':', 1540), ('even', 1381), ('would', 1185), ('good', 1126), ('time', 1111), ('!', 1056), ('get', 1039), ('bad', 1019)]


Removing Stop words in Positive Files

In [21]:
filtered_sentence1 = [] 

In [22]:
for w in pos_text:  
    if w not in stop_words:  
        filtered_sentence1.append(w) 

Top 20 frequent tokens in Positive files :

In [23]:
frequency1 = {}

for token in filtered_sentence1:
  if token in frequency1: frequency1[token] +=1
  else : frequency1[token] = 1

In [24]:
t = sorted(frequency1.items(), key=lambda x:-x[1])[:20]

In [94]:
print("Top 20 frequent tokens in Positive files are :")
print("\n")
print(t)

Top 20 frequent tokens in Positive files are :


[(',', 42448), ('.', 33714), ("'s", 9473), ('``', 8494), (')', 6039), ('(', 6014), ('film', 5186), ('one', 2943), ("n't", 2775), ('movie', 2497), ('like', 1713), ('?', 1570), (':', 1502), ('story', 1231), ('also', 1200), ('good', 1190), ('even', 1175), ('time', 1171), ('would', 1079), ('character', 1067)]


What kind of things do you notice are
different between the two sets ?

a) The counts for the same tokens in both files are different, eg: movie token has count 2497 in positive files where as a count of 3174 in negative files.

b) The order of frequency of tokens are different for both positive and negative files.

c) There are also changes in tokens while listing out the most frequent tokens in both positive and negative files.

 **2)Using the code from previous lectures, build 3 polarity classifiers using the following
parameters (20 points). Note: just train the models.
a) For training: use 50% of the positive dataset and 70% of the negative dataset. For
your model use: NaiveBayes with the TF-IDF vectorizer.
b) For training: use 50% of the negative dataset and 70% of the positive dataset. For
your model use: NaiveBayes with the TF-IDF vectorizer.
c) For training: use 25% of the negative dataset and 25% of the positive dataset. For
your model use: SVM with the TF-IDF vectorizer.**


In [26]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.pipeline import make_pipeline
from sklearn.svm import SVC

from sklearn.metrics import accuracy_score
from sklearn.metrics import precision_score
from sklearn.metrics import recall_score
from sklearn.metrics import f1_score
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_squared_error

In [27]:
# a) NaiveBayes and TF-IDF vectorizer
w = []
co = 0
g = []

#counting number of files in positive
for h in txt :
  co +=1

n_co = int(.50*co)

#taking 50%
w = d[:n_co]
g = [1]*n_co

In [28]:
c = 0
k = []
e = []

#counting number of files in negative
for h in txt1 :
  c+=1

n_c = int(.70*c) 

#taking 70%
k = da[:n_c]
e = [0]*n_c

In [29]:
data = w+k
label = g+e

In [30]:
len(data)
len(label)

1200

In [31]:
model1 = make_pipeline(TfidfVectorizer(), MultinomialNB())
model1.fit(data,label) 

Pipeline(memory=None,
         steps=[('tfidfvectorizer',
                 TfidfVectorizer(analyzer='word', binary=False,
                                 decode_error='strict',
                                 dtype=<class 'numpy.float64'>,
                                 encoding='utf-8', input='content',
                                 lowercase=True, max_df=1.0, max_features=None,
                                 min_df=1, ngram_range=(1, 1), norm='l2',
                                 preprocessor=None, smooth_idf=True,
                                 stop_words=None, strip_accents=None,
                                 sublinear_tf=False,
                                 token_pattern='(?u)\\b\\w\\w+\\b',
                                 tokenizer=None, use_idf=True,
                                 vocabulary=None)),
                ('multinomialnb',
                 MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True))],
         verbose=False)

In [32]:
# b) NaiveBayes and TF-IDF vectorizer
wi = []
co = 0
gi = []

#counting number of files in positvie
for h in txt :
  co +=1

#taking 70%
n_co = int(.70*co)

wi = d[:n_co]
gi = [1]*n_co

In [33]:
len(wi)

700

In [34]:
c = 0
ki = []
ei = []

#counting number of files in neagtive
for h in txt1 :
  c+=1

n_c = int(.50*c) 

#taking 50%
ki = da[:n_c]
ei = [0]*n_c

In [35]:
data1 = wi+ki
label1 = gi+ei

In [36]:
model2 = make_pipeline(TfidfVectorizer(), MultinomialNB())
model2.fit(data1,label1) 

Pipeline(memory=None,
         steps=[('tfidfvectorizer',
                 TfidfVectorizer(analyzer='word', binary=False,
                                 decode_error='strict',
                                 dtype=<class 'numpy.float64'>,
                                 encoding='utf-8', input='content',
                                 lowercase=True, max_df=1.0, max_features=None,
                                 min_df=1, ngram_range=(1, 1), norm='l2',
                                 preprocessor=None, smooth_idf=True,
                                 stop_words=None, strip_accents=None,
                                 sublinear_tf=False,
                                 token_pattern='(?u)\\b\\w\\w+\\b',
                                 tokenizer=None, use_idf=True,
                                 vocabulary=None)),
                ('multinomialnb',
                 MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True))],
         verbose=False)

In [37]:
# c) SVM with the TF-IDF vectorizer

wii = []
co = 0
gii = []

#counting number of files in positvie
for h in txt :
  co +=1

n_co = int(.25*co)

#taking 25%
wii = d[:n_co]
gii = [1]*n_co

In [38]:
c = 0
kii = []
eii = []

#counting number of files in negative
for h in txt1 :
  c+=1

n_c = int(.25*c) 

#taking 25%
kii = da[:n_c]
eii = [0]*n_c

In [39]:
len(kii)

250

In [40]:
data2 = wii+kii
label2 = gii+eii

In [41]:
model3 = make_pipeline(TfidfVectorizer(), SVC())
model3.fit(data2,label2) 

Pipeline(memory=None,
         steps=[('tfidfvectorizer',
                 TfidfVectorizer(analyzer='word', binary=False,
                                 decode_error='strict',
                                 dtype=<class 'numpy.float64'>,
                                 encoding='utf-8', input='content',
                                 lowercase=True, max_df=1.0, max_features=None,
                                 min_df=1, ngram_range=(1, 1), norm='l2',
                                 preprocessor=None, smooth_idf=True,
                                 stop_words=None, strip_accents=None,
                                 sublinear_tf=False,
                                 token_pattern='(?u)\\b\\w\\w+\\b',
                                 tokenizer=None, use_idf=True,
                                 vocabulary=None)),
                ('svc',
                 SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None,
                     coef0=0.0, decision_function_shape=

**3) Using the models from question 2, evaluate them on their individual rest of the dataset.
This is, for a) 50% positive and 30% negative, for b) 50% negative and 30% positive, and
for c) 75% negative and 75% positive. Calculate and show ONLY the following metrics for
each model: Accuracy, Precision, Recall, Macro F1-score.** (15 points).

In [42]:
# a) NaiveBayes and TF-IDF vectorizer
we = []
co = 0
ge = []

#counting number of files in positvie
for h in txt :
  co +=1

n_co = int(.50*co)

#taking remaining 50%
we = d[n_co:]
ge = [1]*n_co

In [43]:
c = 0
ke = []
ee = []

#counting number of files in negative
for h in txt1 :
  c+=1

n_c = int(.70*c) 

#taking remaining 30%
ke = da[n_c:]
ee = [0]*300

In [44]:
xtest1 = we+ke
ytest1 = ge+ee

In [45]:
l1=model1.predict(xtest1)

In [46]:
print('Accuracy:',accuracy_score(ytest1,l1))
acc = accuracy_score(ytest1,l1)
print('Precision:', precision_score(ytest1,l1))
pre = precision_score(ytest1,l1)
print('Recall:', recall_score(ytest1,l1))
rec = recall_score(ytest1,l1)
print('F1 Score:', f1_score(l1, ytest1))
f1 = f1_score(l1, ytest1)

Accuracy: 0.37625
Precision: 1.0
Recall: 0.002
F1 Score: 0.003992015968063872


In [47]:
import pandas as pd

df = pd.DataFrame([acc,pre,rec,f1],index=['Accuracy','Precision','Recall','F1 Score'],columns=['Model 1'])
print(df)

            Model 1
Accuracy   0.376250
Precision  1.000000
Recall     0.002000
F1 Score   0.003992


In [48]:
# b) NaiveBayes and TF-IDF vectorizer
wk = []
co = 0
gk = []

#counting number of files in positive
for h in txt :
  co +=1

n_co = int(.70*co)

#taking remaining 30%
wk = d[n_co:]
gk = [1]*300

In [49]:
c = 0
kq = []
eq = []

#counting number of files in negative
for h in txt1 :
  c+=1

n_c = int(.50*c)

#taking remaining 50%
kq = da[n_c:]
eq = [0]*n_c

In [50]:
xtest2 = wk+kq
ytest2 = gk+eq

In [51]:
l2=model2.predict(xtest2)

In [52]:
print('Accuracy:',accuracy_score(ytest2,l2))
acc = accuracy_score(ytest2,l2)
print('Precision:', precision_score(ytest2,l2))
pre = precision_score(ytest2,l2)
print('Recall:', recall_score(ytest2,l2))
rec = recall_score(ytest2,l2)
print('F1 Score:', f1_score(l2, ytest2,average='macro'))
f1 = f1_score(l2, ytest2,average='macro')

Accuracy: 0.375
Precision: 0.375
Recall: 1.0
F1 Score: 0.2727272727272727


In [53]:
Model2 = [acc,pre,rec,f1]
df['Model 2'] = Model2
print(df)

            Model 1   Model 2
Accuracy   0.376250  0.375000
Precision  1.000000  0.375000
Recall     0.002000  1.000000
F1 Score   0.003992  0.272727


In [54]:
# c) SVM with the TF-IDF vectorizer
hi = []
co = 0
ai = []

#counting number of files in postive
for h in txt :
  co +=1

n_co = int(.25*co)

#taking remaining 75%
hi = d[n_co:]
ai = [1]*750

In [55]:
c = 0
ji = []
li = []

#counting number of files in negative
for h in txt1 :
  c+=1

n_c = int(.25*c)

#taking remaining 75%
ji = da[n_c:]
li = [0]*750

In [56]:
xtest3 = hi+ji
ytest3 = ai+li

In [57]:
l3 = model3.predict(xtest3)

In [58]:
print('Accuracy:',accuracy_score(ytest3,l3))
acc = accuracy_score(ytest3,l3)
print('Precision:', precision_score(ytest3,l3))
pre = precision_score(ytest3,l3)
print('Recall:', recall_score(ytest3,l3))
rec = recall_score(ytest3,l3)
print('F1 Score:', f1_score(l3, ytest3,average='macro'))
f1 = f1_score(l3, ytest3,average='macro')

Accuracy: 0.76
Precision: 0.75
Recall: 0.78
F1 Score: 0.7599039615846339


In [59]:
Model3 = [acc,pre,rec,f1]
df['Model 3'] = Model3
print(df)

            Model 1   Model 2   Model 3
Accuracy   0.376250  0.375000  0.760000
Precision  1.000000  0.375000  0.750000
Recall     0.002000  1.000000  0.780000
F1 Score   0.003992  0.272727  0.759904


**4)Using the model performance metrics from question 3, answer the following
questions. Please provide logical and intuitive rationale for your answers, simple
answers like: because it has the best score, will not be sufficient. (40 points):**


a) What is the best performing model?

Ans : SVM with TF-IDF is the best performing model.

b) Why do you think this is the best performing model?

Ans : It has balanced training set for positive and negative datasets.

c) How does class imbalance play in determining polarity?

Ans : The accuracy for first Naive Bayes and second Naive Bayes models are less. Meaning their classifications are not fully correct. Also, model 1 has a precision = 1 and model 2 has recall = 1, which is not ideal.


d) Do you think either more data or a better model is a better approach for this kind of task?

Ans : No, more data or a better model does not form a a better approach for this kind of task. One needs to make sure of overfitting and underfitting scenarios to select more data or better model. Best case is to find a optimum way with a mix of both.

**5) Using NLTK and VADER, calculate the sentiment score for all documents in the
positive polarity. Calculate the polarity threshold needed (and reasonable) to have the
majority of the document labels match. Do the same for the negative class. Provide the
threshold needed, the reason why you think this threshold is reasonable, and the
accuracy percentage (how many documents are correctly labeled using this threshold).**
(45 points):


In [95]:
pip install vaderSentiment

     |████████████████████████████████| 133kB 8.4MB/s 


In [96]:
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer

Calculating sentiment score for positive polarity files :

In [115]:
poslist = []

def sentiment_scores(text): 
   
    sid_obj = SentimentIntensityAnalyzer() 
  
    sentiment_dict = sid_obj.polarity_scores(text) 
      
    print("Overall sentiment dictionary is : ", sentiment_dict) 
    print("file was rated as ", sentiment_dict['neg']*100, "% Negative") 
    print("file was rated as ", sentiment_dict['neu']*100, "% Neutral") 
    print("file was rated as ", sentiment_dict['pos']*100, "% Positive") 
  
    print("File Overall Rated As", end = " ") 

    if sentiment_dict['compound'] >= 0.05 : 
      print("Positive") 
      poslist.append(sentiment_dict['pos'])

    elif sentiment_dict['compound'] > -0.05 and sentiment_dict['compound'] < 0.05 : 
      print("Neutral") 

    elif sentiment_dict['compound'] <= - 0.05 : 
      print("Negative") 

    print("\n")  

In [116]:
for f in txt :
  print("File : ",f)
  m = open(os.path.join(path,f), 'r').read()
  sentiment_scores(m)

Streaming output truncated to the last 5000 lines.
File :  cv960_29007.txt
Overall sentiment dictionary is :  {'neg': 0.105, 'neu': 0.808, 'pos': 0.087, 'compound': -0.9783}
file was rated as  10.5 % Negative
file was rated as  80.80000000000001 % Neutral
file was rated as  8.7 % Positive
File Overall Rated As Negative


File :  cv885_12318.txt
Overall sentiment dictionary is :  {'neg': 0.05, 'neu': 0.806, 'pos': 0.144, 'compound': 0.9963}
file was rated as  5.0 % Negative
file was rated as  80.60000000000001 % Neutral
file was rated as  14.399999999999999 % Positive
File Overall Rated As Positive


File :  cv045_23923.txt
Overall sentiment dictionary is :  {'neg': 0.039, 'neu': 0.851, 'pos': 0.111, 'compound': 0.997}
file was rated as  3.9 % Negative
file was rated as  85.1 % Neutral
file was rated as  11.1 % Positive
File Overall Rated As Positive


File :  cv994_12270.txt
Overall sentiment dictionary is :  {'neg': 0.067, 'neu': 0.796, 'pos': 0.137, 'compound': 0.9922}
file was rated

In [117]:
def Average(list1): 
    return sum(list1) / len(list1)

In [118]:
average = Average(poslist)
print("The threshold required for positive files are : ",average)

The threshold required for positive files are :  0.13817674970344013


Calculating Sentiment score for negative polarity files :

In [125]:
neglist = []

def sentiment_scores1(text): 
   
    sid_obj1 = SentimentIntensityAnalyzer() 
  
    sentiment_dict1 = sid_obj1.polarity_scores(text) 
      
    print("Overall sentiment dictionary is : ", sentiment_dict1) 
    print("file was rated as ", sentiment_dict1['neg']*100, "% Negative") 
    print("file was rated as ", sentiment_dict1['neu']*100, "% Neutral") 
    print("file was rated as ", sentiment_dict1['pos']*100, "% Positive") 
  
    print("File Overall Rated As", end = " ") 

    if sentiment_dict1['compound'] >= 0.05 : 
      print("Positive")     

    elif sentiment_dict1['compound'] > -0.05 and sentiment_dict1['compound'] < 0.05 : 
      print("Neutral") 

    elif sentiment_dict1['compound'] <= - 0.05 : 
      print("Negative") 
      neglist.append(sentiment_dict1['neg'])

    print("\n")  

In [126]:
for h in txt1 :
  print("File : ", h)
  n = open(os.path.join(path1,h), 'r').read()
  sentiment_scores1(n)

Streaming output truncated to the last 5000 lines.
File :  cv951_11816.txt
Overall sentiment dictionary is :  {'neg': 0.114, 'neu': 0.747, 'pos': 0.14, 'compound': 0.9753}
file was rated as  11.4 % Negative
file was rated as  74.7 % Neutral
file was rated as  14.000000000000002 % Positive
File Overall Rated As Positive


File :  cv173_4295.txt
Overall sentiment dictionary is :  {'neg': 0.167, 'neu': 0.739, 'pos': 0.094, 'compound': -0.9872}
file was rated as  16.7 % Negative
file was rated as  73.9 % Neutral
file was rated as  9.4 % Positive
File Overall Rated As Negative


File :  cv342_20917.txt
Overall sentiment dictionary is :  {'neg': 0.122, 'neu': 0.706, 'pos': 0.171, 'compound': 0.9853}
file was rated as  12.2 % Negative
file was rated as  70.6 % Neutral
file was rated as  17.1 % Positive
File Overall Rated As Positive


File :  cv924_29397.txt
Overall sentiment dictionary is :  {'neg': 0.077, 'neu': 0.81, 'pos': 0.114, 'compound': 0.9758}
file was rated as  7.7 % Negative
file 

In [127]:
average = Average(neglist)
print("The threshold required for negative files are : ",average)

The threshold required for negative files are :  0.11913425925925925


**Bonus (40 points): Repeat questions 2,3 and 4 removing all stopwords. Answer the
following questions: Did this change the results in any way? Why do you think so?**

In [60]:
# Removing stop words from Positive files
final_data = []

for f in txt :
  b = open(os.path.join(path,f), 'r').read()
  out_string = ''
  tok = word_tokenize(b)
  for w in tok:  
    if w not in stop_words:  
        out_string += w + ' '
  final_data.append(out_string)

In [61]:
# Removing stop words from negative files
final_data1 = []

for h in txt1 :
  j = open(os.path.join(path1,h), 'r').read()
  out_string1 = ''
  tok = word_tokenize(j)
  for w in tok:  
    if w not in stop_words:  
        out_string1 += w + ' '
  final_data1.append(out_string1)

Making models

In [62]:
# a) NaiveBayes and TF-IDF vectorizer
w = []
co = 0
g = []

#counting number of files in postive
for h in txt :
  co +=1

n_co = int(.50*co)

#taking 50%
w = final_data[:n_co]
g = [1]*n_co

In [63]:
c = 0
k = []
e = []

#counting number of files in neagtive
for h in txt1 :
  c+=1

n_c = int(.70*c) 

#taking 70%
k = final_data1[:n_c]
e = [0]*n_c

In [64]:
q1 = w+k
z1 = g+e

In [65]:
m1 = make_pipeline(TfidfVectorizer(), MultinomialNB())
m1.fit(q1,z1) 

Pipeline(memory=None,
         steps=[('tfidfvectorizer',
                 TfidfVectorizer(analyzer='word', binary=False,
                                 decode_error='strict',
                                 dtype=<class 'numpy.float64'>,
                                 encoding='utf-8', input='content',
                                 lowercase=True, max_df=1.0, max_features=None,
                                 min_df=1, ngram_range=(1, 1), norm='l2',
                                 preprocessor=None, smooth_idf=True,
                                 stop_words=None, strip_accents=None,
                                 sublinear_tf=False,
                                 token_pattern='(?u)\\b\\w\\w+\\b',
                                 tokenizer=None, use_idf=True,
                                 vocabulary=None)),
                ('multinomialnb',
                 MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True))],
         verbose=False)

In [66]:
# b) NaiveBayes and TF-IDF vectorizer
wi = []
co = 0
gi = []

#counting number of files in postive
for h in txt :
  co +=1

n_co = int(.70*co)

#taking 70%
wi = final_data[:n_co]
gi = [1]*n_co

In [67]:
c = 0
ki = []
ei = []

#counting number of files in negative
for h in txt1 :
  c+=1

n_c = int(.50*c)

#taking 50%
ki = final_data1[:n_c]
ei = [0]*n_c

In [68]:
q2 = wi+ki
z2 = gi+ei

In [69]:
m2 = make_pipeline(TfidfVectorizer(), MultinomialNB())
m2.fit(q2,z2)

Pipeline(memory=None,
         steps=[('tfidfvectorizer',
                 TfidfVectorizer(analyzer='word', binary=False,
                                 decode_error='strict',
                                 dtype=<class 'numpy.float64'>,
                                 encoding='utf-8', input='content',
                                 lowercase=True, max_df=1.0, max_features=None,
                                 min_df=1, ngram_range=(1, 1), norm='l2',
                                 preprocessor=None, smooth_idf=True,
                                 stop_words=None, strip_accents=None,
                                 sublinear_tf=False,
                                 token_pattern='(?u)\\b\\w\\w+\\b',
                                 tokenizer=None, use_idf=True,
                                 vocabulary=None)),
                ('multinomialnb',
                 MultinomialNB(alpha=1.0, class_prior=None, fit_prior=True))],
         verbose=False)

In [70]:
# c) SVM with the TF-IDF vectorizer

wii = []
co = 0
gii = []

#counting number of files in postive
for h in txt :
  co +=1

n_co = int(.25*co)

#taking 25%
wii = final_data[:n_co]
gii = [1]*n_co

In [71]:
c = 0
kii = []
eii = []

#counting number of files in negative
for h in txt1 :
  c+=1

n_c = int(.25*c)

#taking 25%
kii = final_data1[:n_c]
eii = [0]*n_c

In [72]:
q3 = wii+kii
z3 = gii+eii

In [73]:
m3 = make_pipeline(TfidfVectorizer(), SVC(gamma='auto'))
m3.fit(q3,z3) 

Pipeline(memory=None,
         steps=[('tfidfvectorizer',
                 TfidfVectorizer(analyzer='word', binary=False,
                                 decode_error='strict',
                                 dtype=<class 'numpy.float64'>,
                                 encoding='utf-8', input='content',
                                 lowercase=True, max_df=1.0, max_features=None,
                                 min_df=1, ngram_range=(1, 1), norm='l2',
                                 preprocessor=None, smooth_idf=True,
                                 stop_words=None, strip_accents=None,
                                 sublinear_tf=False,
                                 token_pattern='(?u)\\b\\w\\w+\\b',
                                 tokenizer=None, use_idf=True,
                                 vocabulary=None)),
                ('svc',
                 SVC(C=1.0, break_ties=False, cache_size=200, class_weight=None,
                     coef0=0.0, decision_function_shape=

**Testing Models :**

In [74]:
# a) NaiveBayes and TF-IDF vectorizer
we = []
co = 0
ge = []

#counting number of files in postive
for h in txt :
  co +=1

n_co = int(.50*co)

#taking remaining 50%
we = final_data[n_co:]
ge = [1]*n_co

In [75]:
c = 0
ke = []
ee = []

#counting number of files in negative
for h in txt1 :
  c+=1

n_c = int(.70*c)

#taking remaining 30%
ke = final_data1[n_c:]
ee = [0]*300

In [76]:
xt1 = we+ke
yt1 = ge+ee

In [77]:
la1=m1.predict(xt1)

In [78]:
print('Accuracy:',accuracy_score(yt1,la1))
acc = accuracy_score(yt1,la1)
print('Precision:', precision_score(yt1,la1))
pre = precision_score(yt1,la1)
print('Recall:', recall_score(yt1,la1))
rec = recall_score(yt1,la1)
print('F1 Score:', f1_score(la1, yt1, average='macro'))
f1 = f1_score(la1, yt1, average='macro')

Accuracy: 0.395
Precision: 1.0
Recall: 0.032
F1 Score: 0.30776051946565974


In [79]:
df = pd.DataFrame([acc,pre,rec,f1],index=['Accuracy','Precision','Recall','F1 Score'],columns=['Model 1'])
print(df)

            Model 1
Accuracy   0.395000
Precision  1.000000
Recall     0.032000
F1 Score   0.307761


In [80]:
# b) NaiveBayes and TF-IDF vectorizer
wk = []
co = 0
gk = []

#counting number of files in postive
for h in txt :
  co +=1

n_co = int(.70*co)

#taking remaining 30%
wk = final_data[n_co:]
gk = [1]*300

In [81]:
c = 0
kq = []
eq = []

#counting number of files in negative
for h in txt1 :
  c+=1

n_c = int(.50*c)

#taking remaining 50%
kq = final_data1[n_c:]
eq = [0]*n_c

In [82]:
xt2 = wk+kq
yt2 = gk+eq

In [83]:
la2=m2.predict(xt2)

In [84]:
print('Accuracy:',accuracy_score(yt2,la2))
acc = accuracy_score(yt2,la2)
print('Precision:', precision_score(yt2,la2))
pre = precision_score(yt2,la2)
print('Recall:', recall_score(yt2,la2))
rec = recall_score(yt2,la2)
print('F1 Score:', f1_score(la2, yt2, average='macro'))
f1 = f1_score(la2, yt2, average='macro')

Accuracy: 0.37875
Precision: 0.37641154328732745
Recall: 1.0
F1 Score: 0.279437323189396


In [85]:
Model2 = [acc,pre,rec,f1]
df['Model 2'] = Model2
print(df)

            Model 1   Model 2
Accuracy   0.395000  0.378750
Precision  1.000000  0.376412
Recall     0.032000  1.000000
F1 Score   0.307761  0.279437


In [86]:
# c) SVM with the TF-IDF vectorizer
hi = []
co = 0
ai = []

#counting number of files in postive
for h in txt :
  co +=1

n_co = int(.25*co)

#taking remaining 75%
hi = final_data[n_co:]
ai = [1]*750

In [87]:
c = 0
ji = []
li = []

#counting number of files in negative
for h in txt1 :
  c+=1

n_c = int(.25*c)

#taking remaining 75%
ji = final_data1[n_c:]
li = [0]*750

In [88]:
xt3 = hi+ji
yt3 = ai+li

In [89]:
la3 = m3.predict(xt3)

In [90]:
print('Accuracy:',accuracy_score(yt3,la3))
acc = accuracy_score(yt3,la3)
print('Precision:', precision_score(yt3,la3))
pre = precision_score(yt3,la3)
print('Recall:', recall_score(yt3,la3))
rec = recall_score(yt3,la3)
print('F1 Score:', f1_score(la3, yt3, average='macro'))
f1 = f1_score(la3, yt3, average='macro')

Accuracy: 0.7713333333333333
Precision: 0.7185821697099892
Recall: 0.892
F1 Score: 0.7679546499046788


In [91]:
Model3 = [acc,pre,rec,f1]
df['Model 3'] = Model3
print(df)

            Model 1   Model 2   Model 3
Accuracy   0.395000  0.378750  0.771333
Precision  1.000000  0.376412  0.718582
Recall     0.032000  1.000000  0.892000
F1 Score   0.307761  0.279437  0.767955


Removal of stopwords does'nt bring much difference in the performance of the models.Because stopwords doesnt contribute much in deciding the polarity of the files as well as in calculating accuracy, precision, recall and F1 score
Here, SVM is still the best performing model.
